frontend/packages/metal3-plugin/src/components/host.tsx

```diff
@@ -124,81 +101,96 @@ const HostsTableRow: React.FC<HostsTableRowProps> = ({
+        />
+      </TableData>
     </TableRow>
   );
 };
 
-type HostListProps = {
-  data: K8sResourceKind[];
-  resources: {
-    machines: FirehoseResult<MachineKind[]>;
-    nodes: FirehoseResult<NodeKind[]>;
-  };
-} & React.ComponentProps<typeof Table>;
-
-const HostList: React.FC<HostListProps> = (props) => {
-  const {
-    resources: { machines, nodes },
-    ...tableProps
-  } = props;
-  return (
-    <Table
-      {...tableProps}
-      aria-label="Baremetal Hosts"
-      Header={HostsTableHeader}
-      Row={HostsTableRow}
-      customData={{
-        machines: machines.data || [],
-        nodes: nodes.data || [],
-      }}
-    />
-  );
-};
-
-type BaremetalHostsPageProps = {
-  namespace: string;
-};
-
-export const BaremetalHostsPage: React.FC<BaremetalHostsPageProps> = (props) => {
-  const hostsResource = {
-    kind: referenceForModel(BaremetalHostModel),
-    namespaced: true,
-    prop: 'hosts',
-  };
-  const machinesResource = {
-    kind: referenceForModel(MachineModel),
-    namespaced: true,
-    prop: 'machines',
-  };
-  const nodesResource = {
-    kind: NodeModel.kind,
-    namespaced: false,
-    prop: 'nodes',
-  };
-
-  const flatten = (resources) => {
+const HostList: React.FC<React.ComponentProps<typeof Table> & HostListProps> = (props) => (
+  <Table {...props} aria-label="Baremetal Hosts" Header={HostsTableHeader} Row={HostsTableRow} />
+);
+
+const BaremetalHostsPage: React.FC<BaremetalHostsPageProps> = (props) => {
+  const resources: FirehoseResource[] = [
+    {
+      kind: referenceForModel(BaremetalHostModel),
+      namespaced: true,
+      prop: 'hosts',
+    },
+    {
+      kind: referenceForModel(MachineModel),
+      namespaced: true,
+      prop: 'machines',
+    },
+    {
+      kind: NodeModel.kind,
+      namespaced: false,
+      prop: 'nodes',
+    },
+  ];
+
+  if (props.hasNodeMaintenanceCapability) {
+    resources.push({
+      kind: referenceForModel(NodeMaintenanceModel),
+      namespaced: false,
+      isList: true,
+      prop: 'nodeMaintenances',
+      optional: true,
+    });
+  }
+
+  const flatten = (fResources) => {
     // TODO(jtomasek): Remove loaded check once ListPageWrapper_ is updated to call flatten only
     // when resources are loaded
-    const loaded = _.every(resources, (resource) =>
+    const loaded = _.every(fResources, (resource) =>
       resource.optional ? resource.loaded || !_.isEmpty(resource.loadError) : resource.loaded,
     );
     const {
       hosts: { data: hostsData },
-    } = resources;
-
-    return loaded ? hostsData : [];
+      machines: { data: machinesData },
+      nodes: { data: nodesData },
+      nodeMaintenances,
+    } = fResources;
+
+    if (loaded) {
+      const nodeMaintenanceLookup = createLookup(nodeMaintenances, getNodeMaintenanceNodeName);
+
+      return hostsData.map((host) => {
+        const machine = getHostMachine(host, machinesData);
+        const node = nodesData && getMachineNode(machine, nodesData);
```

Comment:
Again, `getMachineNode` should use a sensible fallback for `nodesData` argument, which in turn simplifies the caller code.